### ISSUE TITLE
fails to unmount directory in non-world-readable directory on NFS filesystem mounted without no_root_squash

### ISSUE THREAD

**Issue by droundy** [jacereda/traced-fs#1]
This is the bug I mentioned before that hits tup.  The 

```
bennet:traced-fs$ killall fs
bennet:traced-fs$ fusermount: failed to chdir to /home/droundy/.tmp/traced-fs: Permission denied
bennet:traced-fs$ mount
...
knightley:/srv/home on /home type nfs4 (rw,relatime,vers=4.0,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=128.193.96.83,local_lock=none,addr=128.193.96.85,_netdev)
fs on /home/droundy/.tmp/traced-fs/traced type fuse.fs (rw,nosuid,nodev,relatime,user_id=1137,group_id=100)
```

The filesystem can be unmounted by making the directory world-readable.  The problem I see here is that there is hint to a user that this is what to do.  The other problem, of course, being that users using NFS often have little control over the configuration of the server, and requiring the directory to be world-readable seems potentially problematic.

**Comment by jacereda** [comment]
I'm trying to reproduce this on a Mac. This is what I tried:

Wrote to /etc/exports something like:

```
/Users -alldirs
```

Launched nfsd:

```
sudo nfsd
```

Mounted the share:

```
mkdir Users
mount -t nfs localhost:/Users 
```

Changed to the sources directory in the share and launched the fs:

```
cd Users/jacereda/src/traced-fs
./fs &
```

Checked that it works and then tried to kill the fs:

```
killall fs
```

These steps work fine, am I missing something? Am I missing some nfs option? Maybe it's linux-specific?

**Comment by droundy** [comment]
I don't know how NFS works on the mac.  Looking at fs.c, I can see that the bug (if you like) is in fuse_main, which would certainly have a different implementation on the mac.

On linux, the umount system call requires root privileges, so the call to umount(mpt) will always fail, but I guess that doesn't cause trouble because you don't check for errors, and having the mount point multiply mounted shouldn't cause a failure.  Anyhow, I expect this is the distinction, on linux umount is a privileged system call, which means that fuse needs to use an SUID executable to unmount the directory when it is killed.

I doubt this is something you can fix, but do think it's worth being aware of.  And it's a reason why I won't rely on a FUSE mount under linux.

**Comment by jacereda** [comment]
Would shelling out to `fusermount` instead of calling `umount` fix the problem maybe?

**Comment by droundy** [comment]
No, fusermount also fails, for the same reason.  In fact, there is no way to unmount (that I know of) without changing the directory to be world-readable.

**Comment by droundy** [comment]
I should correct myself, calling fusermount instead of umount at the beginning would fix the bug for the case where root has read permissions on the directory, which is the common case.

**Comment by jacereda** [comment]
So, would something like this do the trick?

https://github.com/jacereda/traced-fs/commit/8c6600aefbfae2b4535c06abcb31852937deca3e

**Comment by droundy** [comment]
Something like that ought to work, although you really ought to use fork+exec, so as to correctly handle the case where the file path has embedded quotes.

I guess for now mpl is always "traced", but I presume the plan is to create more flexibility at some stage.

**Comment by BrianGraysonSiV** [comment]
In case anyone still is wondering about this, it looks like it's been diagnosed and fixed: https://github.com/libfuse/libfuse/commit/774eb66fc1eb006b601f3890958f02e7b90575ad
